File: backend-elixir/test/unit/external_auth_test.exs
defmodule BackendElixir.Auth.ExternalAuthTest do
  @moduledoc """
  外部公開API専用の認証チェック(ExternalAuth.require_external_client/3)を、実DBを使わずに
  モックJWKSサーバー(Keycloak相当)とローカルHMACのDispatcher経由で検証する
  """
  use ExUnit.Case, async: false

  alias BackendElixir.Auth.{Dispatcher, ExternalAuth, HmacVerifier, JwksCacheServer, JwksVerifier}
  alias BackendElixir.TestSupport.{MockJwksServer, TestTokenHelper}

  @keycloak_issuer "http://localhost:8082/realms/training"
  @audience "backend"
  @external_api_client_id "external-api-client"
  @hmac_secret "test-external-hmac-secret"

  setup do
    {:ok, server_info} = MockJwksServer.start()
    MockJwksServer.reset_keys()
    {:ok, cache_pid} = JwksCacheServer.start_link(jwks_url: server_info.jwks_url)

    dispatcher =
      Dispatcher.new()
      |> Dispatcher.register(
        Dispatcher.local_hmac_issuer(),
        HmacVerifier,
        HmacVerifier.new(@hmac_secret, Dispatcher.local_hmac_issuer(), @audience)
      )
      |> Dispatcher.register(@keycloak_issuer, JwksVerifier, JwksVerifier.new(cache_pid, @keycloak_issuer, @audience))

    on_exit(fn -> MockJwksServer.stop(server_info) end)

    %{dispatcher: dispatcher}
  end

  defp add_key(kid) do
    private_key = TestTokenHelper.generate_rsa_key_pair()
    MockJwksServer.add_key(kid, TestTokenHelper.public_jwk_fields(private_key))
    private_key
  end

  test "accepts a Keycloak-style token with matching azp", %{dispatcher: dispatcher} do
    private_key = add_key("kid-ext-1")
    token = TestTokenHelper.make_rsa_token_with_azp(private_key, "kid-ext-1", @keycloak_issuer, @audience, "user-sub", @external_api_client_id, 3600)

    assert {:ok, claims} = ExternalAuth.require_external_client(dispatcher, "Bearer " <> token, @external_api_client_id)
    assert claims.azp == @external_api_client_id
  end

  test "rejects a valid token with a mismatched azp", %{dispatcher: dispatcher} do
    private_key = add_key("kid-ext-2")
    token = TestTokenHelper.make_rsa_token_with_azp(private_key, "kid-ext-2", @keycloak_issuer, @audience, "user-sub", "some-other-client", 3600)

    assert {:error, %{kind: :unauthenticated}} =
             ExternalAuth.require_external_client(dispatcher, "Bearer " <> token, @external_api_client_id)
  end

  test "rejects a valid token with no azp claim at all", %{dispatcher: dispatcher} do
    private_key = add_key("kid-ext-3")
    token = TestTokenHelper.make_rsa_token(private_key, "kid-ext-3", @keycloak_issuer, @audience, "user-sub", 3600)

    assert {:error, %{kind: :unauthenticated}} =
             ExternalAuth.require_external_client(dispatcher, "Bearer " <> token, @external_api_client_id)
  end

  test "rejects a correctly-signed local HMAC token even with a matching azp claim", %{dispatcher: dispatcher} do
    # ローカル発行issuerは内部REST/gRPC専用であり、署名検証自体が正しく通っても
    # 外部公開APIでは受け付けない(CONTRACT.mdセクション11)
    token =
      TestTokenHelper.make_hmac_token(@hmac_secret, Dispatcher.local_hmac_issuer(), @audience, "1", 3600)

    assert {:error, %{kind: :unauthenticated}} =
             ExternalAuth.require_external_client(dispatcher, "Bearer " <> token, @external_api_client_id)
  end

  test "rejects when the authorization header is missing", %{dispatcher: dispatcher} do
    assert {:error, %{kind: :unauthenticated}} = ExternalAuth.require_external_client(dispatcher, nil, @external_api_client_id)
  end
end

File: backend-js-ts-express/tests/external_auth.test.ts
// backend-js-express/tests/external_auth.test.jsの型付き移植。テスト内容は変更していない。
// backend-java/ExternalAuthTest.java・backend-python/test_external_auth.pyの移植でもある。
// 外部公開API(CONTRACT.mdセクション11)のauthenticate()、
// (1) KeycloakのJWKSで署名検証、(2) azpクレームが期待クライアントIDと一致するか、を検証する。
// DB・実サーバーは不要(モックJWKSサーバーのみ、jwks_verifier.test.tsと同じ設計)

import test from 'node:test';
import assert from 'node:assert/strict';
import type { Request } from 'express';
import { authenticate } from '../src/external/authenticate';
import { JwksVerifier, LOCAL_HMAC_ISSUER, LOCAL_RSA_ISSUER } from '../src/auth/jwt';
import type { ExternalState } from '../src/types';
import { createMockJwksServer } from './support/mock_jwks_server';
import { generateRsaKeyPair, makeRsaToken, makeHmacToken } from './support/test_token_helper';
import type { RestError } from '../src/error';

const KEYCLOAK_ISSUER = 'https://keycloak.example';
const EXTERNAL_CLIENT_ID = 'external-api-client';
const HMAC_SECRET = 'test-secret-at-least-32-bytes-long!!';

function fakeReq(authorizationHeader: string | undefined): Request {
  return { get: (name: string) => (name === 'authorization' ? authorizationHeader : undefined) } as unknown as Request;
}

// pool/paginationV2Flagはこのテストの認証チェックでは使われないため、テスト専用の最小限の状態を作る
function externalState(keycloakVerifier: JwksVerifier): ExternalState {
  return { keycloakVerifier, expectedClientId: EXTERNAL_CLIENT_ID } as unknown as ExternalState;
}

function assertRestError(err: unknown, kind: string, status: number): true {
  const e = err as RestError;
  assert.equal(e.kind, kind);
  assert.equal(e.status, status);
  return true;
}

test('accepts a Keycloak token with a matching azp', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const state = externalState(new JwksVerifier(mockJwks.jwksUrl(), KEYCLOAK_ISSUER, 'backend'));
    const token = makeRsaToken(privateKey, 'kid-1', KEYCLOAK_ISSUER, 'backend', 'some-keycloak-sub', 3600, 'RS256', EXTERNAL_CLIENT_ID);

    await assert.doesNotReject(() => authenticate(state, fakeReq(`Bearer ${token}`)));
  } finally {
    await mockJwks.close();
  }
});

test('rejects a missing Authorization header', async () => {
  const state = externalState(new JwksVerifier('http://127.0.0.1:1/jwks', KEYCLOAK_ISSUER, 'backend'));
  await assert.rejects(
    () => authenticate(state, fakeReq(undefined)),
    (err) => assertRestError(err, 'unauthorized', 401),
  );
});

test('rejects a malformed (non-Bearer) Authorization header', async () => {
  const state = externalState(new JwksVerifier('http://127.0.0.1:1/jwks', KEYCLOAK_ISSUER, 'backend'));
  await assert.rejects(
    () => authenticate(state, fakeReq('not-a-bearer-header')),
    (err) => assertRestError(err, 'unauthorized', 401),
  );
});

test('rejects the wrong azp', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const state = externalState(new JwksVerifier(mockJwks.jwksUrl(), KEYCLOAK_ISSUER, 'backend'));
    const token = makeRsaToken(privateKey, 'kid-1', KEYCLOAK_ISSUER, 'backend', 'sub', 3600, 'RS256', 'some-other-client');

    await assert.rejects(
      () => authenticate(state, fakeReq(`Bearer ${token}`)),
      (err) => assertRestError(err, 'client_not_allowed', 403),
    );
  } finally {
    await mockJwks.close();
  }
});

test('rejects a missing azp claim', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const state = externalState(new JwksVerifier(mockJwks.jwksUrl(), KEYCLOAK_ISSUER, 'backend'));
    // azp無し(付与しない)のトークン
    const token = makeRsaToken(privateKey, 'kid-1', KEYCLOAK_ISSUER, 'backend', 'sub', 3600);

    await assert.rejects(
      () => authenticate(state, fakeReq(`Bearer ${token}`)),
      (err) => assertRestError(err, 'client_not_allowed', 403),
    );
  } finally {
    await mockJwks.close();
  }
});

// 【最も見落としやすいケース】ローカルHMAC発行のトークンは、署名自体は別issuer向けの
// verifierでは検証できない(そもそもkeycloakVerifierはKeycloakのissuerしか受け付けない)。
// azpが偶然一致していても、外部公開APIはKeycloak発行(Client Credentials Grant)のみ許可する
test('rejects a locally-issued HMAC token even with a correct azp', async () => {
  const state = externalState(new JwksVerifier('http://127.0.0.1:1/jwks', KEYCLOAK_ISSUER, 'backend'));
  const token = makeHmacToken(HMAC_SECRET, LOCAL_HMAC_ISSUER, 'backend', '1', 3600, 'HS256', EXTERNAL_CLIENT_ID);

  await assert.rejects(
    () => authenticate(state, fakeReq(`Bearer ${token}`)),
    (err) => assertRestError(err, 'invalid_token', 401),
  );
});

// 同様に、ローカルRSA発行(bff自身の/.well-known/jwks.json)のトークンも、
// 正しく署名されRS256/JWKSでかつazpが一致していても、issuerがKeycloakでない限り拒否される
test('rejects a locally-issued RSA token even with a correct azp', async () => {
  const { publicKey, privateKey } = generateRsaKeyPair();
  const mockJwks = await createMockJwksServer();
  try {
    mockJwks.addKey('kid-1', publicKey);
    const state = externalState(new JwksVerifier(mockJwks.jwksUrl(), KEYCLOAK_ISSUER, 'backend'));
    const token = makeRsaToken(privateKey, 'kid-1', LOCAL_RSA_ISSUER, 'backend', '1', 3600, 'RS256', EXTERNAL_CLIENT_ID);

    await assert.rejects(
      () => authenticate(state, fakeReq(`Bearer ${token}`)),
      (err) => assertRestError(err, 'invalid_token', 401),
    );
  } finally {
    await mockJwks.close();
  }
});
